### ISSUE TITLE
PDP 23: Pravega security

### ISSUE THREAD

**Issue by fpj** [pravega/pravega#1950]
**Problem description**
Pravega Design Proposal for security.

See https://github.com/pravega/pravega/wiki/PDP-23:-Pravega-security----encryption-and-Role-Based-Access-Control

**Problem location**
Client, Controller, Segment Store.

**Suggestions for an improvement**
See the PDP for detail.

**Comment by tkaitchuck** [comment]
> All the nodes should authenticate to each other using X.509 certificates. A sample self singed cert will be provided along with Pravega package. Deployers are expected to replace it with a valid root cert from a CA.

Is this really a reasonable expectation? Normally root certs are not used in this way. If we're going to go this route, we're going to have to provide a reasonable way to setup a CA and bootstrap cert distribution on the client. (Even if this isn't a lot of code that we're writing, its non trivial especially for a new user.

**Comment by tkaitchuck** [comment]
In the Authorization section it talks about, creating a protocol using JWT tokens to do the authentication in the service. This seems to be trying to mimic Kerberos. Rather than reinventing this why aren't we simply adopting an established protocol?

**Comment by arvindkandhare** [comment]
This approach is not exactly replicating kerberos. Here is how kerberos fits in in the given apprach:
1. Authorization: This will be pluggable and can use kerberos along with other pluggable implementation. 
2. Delegation token using JWT: These are for validation at the segmentstore. These happens without any RPC and supposed to be fast as it is in the perf critical path. Using kerb here will not help.

**Comment by maddisondavid** [comment]
I didn't see anything in the proposal for Encryption of Data at Rest for Tier-1, do we have any proposals to protect the Bookkeeper data?

**Comment by sarlaccpit** [comment]
The document was written before implementation started so understandably there are some details that got flushed out since then and some that changed the design as things became more concrete.

I'm not sure how high level or detailed the PDP must be and some details belong in implementation details.  For some items, it might be hard to draw that line.

Some comments may or may not belong here (maybe they would be better in README.md files in the top level of the pravega source tree).  For example, some important pieces for implementors of the auth interface need to be described in detail (what the interface is called, which package it's available in, what parameters need to be set on the controller/segment store etc. in order to enable it properly). 

* Section 1.c Regarding encryption at rest, or general handling of bookeeper; I agree with @maddisondavid that there is not much talking about this, in this iteration of the document.    I'll admit I'm also looking at the implementation at the same time as reviewing this PDP.   Perhaps the bookeeper side of things is meant to be handled at a later time.  It may be fine to keep this high level here, but maybe outlining the general strategy of certs would provide enough for this PDP to stand on its own.  For example, would there be a cluster CA for the whole of Pravega, and this CA would sign all certs provided to the various components (zk, controller, segment store etc.)
* Section 1.b: there is a line missing in the diagram I think, which is a direct line between the segment store and the controller. In this PR, https://github.com/pravega/pravega/pull/2175#discussion_r159273484, I had a comment/concern about how this connection is authenticated.  It is an implementation detail but also has a bearing in the general approach of the design wrt authentication between internal components
* Section 3 `cookie` is really a `token` (perhaps cookie was meant in a generic way but it tends to confuse things when calling it different things throughout the document)
* Section 3a:  asymmetric turned out to be symmetric in the implementation
* I feel sections 5.1.a,b,c,d are too lightweight to be useful.  Maybe remove or illustrate?
* 5.2 and 5.3 —> This is where examples, pointers to key interface/classes will be extremely useful for third party implementors (and again it may or not belong in _this_ document but should be called out somewhere).   What parameters are key for controller and segment store and what to set them to, pointers to how to run the password file creator etc.
   * how will be additional auth implementation be picked up?  I'm assuming this is based on this: https://docs.oracle.com/javase/tutorial/ext/basics/spi.html ; some pointers in that direction would be useful
* 5.4. minor: List would be best viewed as a bullet list than flat text.
* Sections 6.2 and 7 are pretty bare.  It could be because this document is meant to evolve in the future?  But in this case

**Comment by arvindkandhare** [comment]
@maddisondavid data at rest is not encrypted by apache bookkeeper. I have added a github issue for this #2224.

**Comment by arvindkandhare** [comment]
@sarlaccpit here is the response for each topic:
1. section 1.c: The approach here is similar to certs for SSL. Pravega just needs a way to access the service account and configure TLS. These will be picked up from Pravega config. The approach is a deployment issue. We can have a default deployment with self-generated certs as an example, in case of a production deployment, it is administrators decision. 
2. section 1.b: We have closed this discussions. Will update the PDP with the decision and more details about this. Will update the image, too.
3. I agree, will replace all the mention of `cookie` with `token`
4. section 3.a: Yes. Need to update the PDP.
5. 5.1 a,b,c,d: These should be bullet points rather than headers. Will update these.
6. Will add more details and a new section about how to implement a custom auth plugin.
7. 5.4: Agree - will reformat this section
8. Section 6.2 - will add more details to the PDP
9. Section 7 - This is just a placeholder and a work in progress. Will update this as we go along.

**Comment by fpj** [comment]
@arvindkandhare some comments:

> Encryption of data at rest (Tier 2)
> The configuration and credentials for doing this should be passed on to Pravega deployment in a secure manner.

How exactly is this supposed to happen? Could you be more concrete, please?

> Encryption of data in flight
> A sample self-signed certificate will be provided along with Pravega package. Deployers are expected to replace it with a valid root cert from a CA.

Could we be more concrete about this step? How is a user expected to obtain a valid cert and use it with Pravega? I understand that most users dealing with security will be familiar with the procedure and there isn't necessarily a single way of doing things, but I'd like us to at least illustrate to make it clear what it is that we are doing.

> Pravega uses service credentials to create and manage data in Tier-1 and tier-2.

I think it is important to point out that this choice limits direct access to data in Tier 2 (or Tier 1 for that matter), unless we start giving the service credentials away, which sounds a bit promiscuous.

> Pravega will implement a pluggable authorization model

As I understand it, the authentication is also part of the pluggable model. I'm asking explicitly because authentication and authorization aren't the same, and we can in principle use multiple combinations of authentication and authorization schemes.

> The token follows the JWT (Jason Web Token) format closely.

Could you justify the use of the JWT format?

> It is signed by a asymmetric key shared between controller and segmentstore.

Could you justify the use of asymmetric keys?

> #### 4. User access to Pravega resources

In this section, please be more concrete and consider illustrating the flow with a figure.

> Tokens are short lived.

What's a good default here?

> Pravega deployments and client implementations should be able to switch between secure and insecure version with a few or no changes.

That's an excellent point: what's the migration path?

> "Pravega-OSS"

This is an odd name, let's call it something that doesn't try to make a distinction between open-source and proprietary, e.g., `DefaultAuthorizer`.

> boolean isTlsEnabled, String tlsCertFile

Do we really need two parameters to enable `TLS`?

Also, I'd rather use config parameters and system properties like we have for zookeeper, unless there is a reason for doing it differently:

https://cwiki.apache.org/confluence/display/ZOOKEEPER/Client-Server+mutual+authentication

Finally, do we need to load a jaas config file?

**Comment by fpj** [comment]
Here is an attack scenario that I don't know how we are dealing with. Say that an attacker brings up a spoofed segment store. If a client connects to such a server, then the server is supposed to authenticate the client using a token. However, since this is a segment store server under the control of an attacker, the server can ignore the token altogether and start receiving append requests, which could lead to unauthorized access to application data.

To deal with this situation, we need to either:
1- Have the client also authenticate the server.
2- Absolutely guarantee that the client will never connect to a spoofed server.

**Comment by sarlaccpit** [comment]
Regarding @fpj  comment above; let me see if my assessment is correct (@arvindkandhare please correct me).   I *think* we are covered because both the controller and the segment store will be started with a cert.    The client will need to trust these certs (or more practically speaking, a CA that signed both of those).

**Comment by arvindkandhare** [comment]
I think this assessment is correct. Server auth will cover the case where an attacker brings in the rogue machine and tries to enter the trusted network.
Other part is where the attacker is able to get his bits on a trusted machine. I think in this scenario he can do more damage than just spoof a user. To do any damage the attacker needs to have service account creds too.

**Comment by EronWright** [comment]
I think @fpj was referring to a man-in-the-middle attack, which is correctly mitigated with server authentication.  There are two variants:
1. The client validates the server certificate against a **public/enterprise** truststore, AND performs hostname verification (by comparing the cert's `dnsName` field against the expected URI).
2. The client validates the server certificate against a **narrow** truststore, skipping hostname verification.

See: https://tools.ietf.org/search/rfc2818#section-3.1 for a good explanation:

> If the client has external information as to the expected identity of
   the server, the hostname check MAY be omitted. (For instance, a
   client may be connecting to a machine whose address and hostname are
   dynamic but the client knows the certificate that the server will
   present.) In such cases, it is important to narrow the scope of
   acceptable certificates as much as possible in order to prevent man in the middle attacks.

**Comment by arvindkandhare** [comment]
> @arvindkandhare some comments:
> 
> Encryption of data at rest (Tier 2)
> The configuration and credentials for doing this should be passed on to Pravega deployment in a secure manner.
> 
> How exactly is this supposed to happen? Could you be more concrete, please?
> 

The implementation of this is still TBD. I was thinking about using hashicorp `vault` for this. Will update the PDP for this.
 
> Encryption of data in flight
> A sample self-signed certificate will be provided along with Pravega package. Deployers are expected to replace it with a valid root cert from a CA.
> 
> Could we be more concrete about this step? How is a user expected to obtain a valid cert and use it with Pravega? I understand that most users dealing with security will be familiar with the procedure and there isn't necessarily a single way of doing things, but I'd like us to at least illustrate to make it clear what it is that we are doing.
> 

Sure. Will add a sessions describing this in more details. This falls in admin guide wiki, but as discussed will update the PDP with this.

> Pravega uses service credentials to create and manage data in Tier-1 and tier-2.
> 
> I think it is important to point out that this choice limits direct access to data in Tier 2 (or Tier 1 for that matter), unless we start giving the service credentials away, which sounds a bit promiscuous.
> 

Depending on tier-2 implementations, admin can give other accounts read access and share those with users directly. Will update PDP with this.


> Pravega will implement a pluggable authorization model
> 
> As I understand it, the authentication is also part of the pluggable model. I'm asking explicitly because authentication and authorization aren't the same, and we can in principle use multiple combinations of authentication and authorization schemes.
> 

Though this is interesting, we have not received any requirement that the authentication and authorization needs to be different. Also the auth is not channel based but call based. Because of this the need for plugin lowers.

> The token follows the JWT (Jason Web Token) format closely.
> 
> Could you justify the use of the JWT format?

Will update the PDP accordingly.

> 
> It is signed by a asymmetric key shared between controller and segmentstore.
> 
> Could you justify the use of asymmetric keys?
> 

Will update the PDP accordingly.

> 4. User access to Pravega resources
> In this section, please be more concrete and consider illustrating the flow with a figure.
> 
> Tokens are short lived.
> 
> What's a good default here?
The ballpark number woould be a couple of minutes. Will update the PDP with the number and justification.

> 
> Pravega deployments and client implementations should be able to switch between secure and insecure version with a few or no changes.
> 
> That's an excellent point: what's the migration path?

Will update the PDP with the path.

> 
> "Pravega-OSS"
> 
> This is an odd name, let's call it something that doesn't try to make a distinction between open-source and proprietary, e.g., DefaultAuthorizer.

Renamed it to "Pravega-Default". Will update the PDP too.
> 
> boolean isTlsEnabled, String tlsCertFile
> 
> Do we really need two parameters to enable TLS?
> 
> Also, I'd rather use config parameters and system properties like we have for zookeeper, unless there is a reason for doing it differently:
> 
> https://cwiki.apache.org/confluence/display/ZOOKEEPER/Client-Server+mutual+authentication
> 

This is a good suggestion. Moving the auth related parameters to system properties.

> Finally, do we need to load a jaas config file?

jaas would come in play for channel based auth. Unfortunately grpc does not support that. We will be doing per call auth using default grpc headers and following the same approach for REST.

**Comment by sarlaccpit** [comment]
Added 👍 .   I think there are sections that will get filled more as we go along, but the general direction is good imo.   This document will become more and more important and it's crucial to keep it up to date as things evolve.  

There are lot of subtopics to discuss, but they are generally more related to implementation so this +1 refers to the design and general direction.      Things like actual settings names, how many flags will be involved to configure TLS etc., probably belong more in the code PRs imho.

Please add segment store/controller connection in the diagram.

In this text:
_"Create an implementation of PravegaCredentials interface: A client selects its auth handler by setting a grpc header with a name "method" the value should match the name of the default implementation. This is done by implementing PravegaCredentials interface and passing it to client calls."_

Is probably "A client selects its auth handler by setting a grpc header with a name "method" the value should match the name of the _**desired custom**_ implementation"

**Comment by andreipaduroiu** [comment]
We need to add some blurb about data stored in the Cache. This means RocksDB. As a refresher, data that still hasn't been transferred to Tier2 will still be stored in the local cache, as well as data pulled out of Tier2 (subject to cache policy). 

Currently, the Cache is implemented by RocksDB which keeps it all in memory, but occasionally offloads to disk if there's memory pressure. The disk data will be cleared on a graceful Segment Container shutdown, but will linger after the process ended in case of a crash. After a Container shuts down (gracefully or not), that data is no longer needed and will not leak into future instances of that container.

We may want to research RocksDB native encryption (ideally we'd only encrypt data that is flushed to disk, but if not, we may have to do the encryption/decryption in our layer. If the latter, special care needs to be given to performance as this lies on the super-hot path for every append, read and tier2 transfer, while holding some locks.

**Comment by arvindkandhare** [comment]
@andreipaduroiu Thanks for the comment. Updated PDP with details about data at rest in BK and rocksdb.

**Comment by arvindkandhare** [comment]
@sarlaccpit Thanks for the +1. updated the PDP as you pointed out.

**Comment by tkaitchuck** [comment]
In the PDP it says:
>Communication between different components need to be secured using the existing mechanisms provided with the components
Communication between BK and ZK
Communication between Pravega SegmentStore and ZK
Communication between Pravega controller and Pravega SegmentStore
Communication between Pravega SegmentStore and ZK

I think we are going to need to go into more depth here. We do need to cover how accounts for each of these systems are created and where the credentials / keys are stored and how they are distributed to new hosts being added to one of these services.

Sections 5.2 and 5.3 are confusing. Are you saying that we will provide a default authentication mechanism and applications will be able to plug in their own? If so please sketch the interface they will need to implement. (Also I don't think we need to name them, a config file could easily just refer to a class.)

In 7.1 the class shouldn't be `PravegaDefaultAuthHandler` but rather `PasswordAuthHandler` or whatever the actual authentication mechanism is. (We don't need to stick the word Pravega into all our classes)

Section 8 needs to be filled out.

There is a second section 6 after section 8.

**Comment by arvindkandhare** [comment]
> I think we are going to need to go into more depth here. We do need to cover how accounts for each of these systems are created and where the credentials / keys are stored and how they are distributed to new hosts being added to one of these services.
> 
> Sections 5.2 and 5.3 are confusing. Are you saying that we will provide a default authentication mechanism and applications will be able to plug in their own? If so please sketch the interface they will need to implement. (Also I don't think we need to name them, a config file could easily just refer to a class.)
> 
> In 7.1 the class shouldn't be PravegaDefaultAuthHandler but rather PasswordAuthHandler or whatever the actual authentication mechanism is. (We don't need to stick the word Pravega into all our classes)

Renamed the class as per the comment.
> 
> Section 8 needs to be filled out.

Added more details about how admin can deploy and enable security.

> There is a second section 6 after section 8.
 Fixed.

**Comment by tkaitchuck** [comment]
In 1.c) 
> It is assumed that this data is secure as the time data spends on disk in these scenarios is very less

This is just not true. We need to have a plan for how to add support for encryption at rest for tier-1.

In 1.d.3)
> This class should support new configurations to support secure communication between segment store and BK.

Which of it's protocols are we going to use Kerberos or TLS? If kerberos, we will need to create documentation for which credentials are going to be used, how to authorize them and how we should tell users to deploy/store them. If TLS mutual authentication we need to provide documentation for how generate the certs in a way that an attacker can't get ahold of their own, and how to store them.

In section 2.3 can you list which wire messages will need to be modified to add a token for the client-server communication? Or if you prefer this can go into section 6.

In section 3.4:
> Controller stores a table containing resource name, ACLs and user id securely

Can you specify how it secures this data?

In the example PravegaAuthHandler in section 7. The last method takes a Object, I don't think this is a good parameter type. Perhaps to avoid complications we should have an init method or something similar that provides this parameter.

In section 8, the user has to do a lot of things here. Perhaps we could have a way for a script to validate the deployment? This also gets in the larger issue of that goes beyond the scope of this PDP: we need to figure out a way to reduce our config burden.

The threat model section still needs to be filled out.
